### ISSUE TITLE
Crash with `Bad file descriptor`

### ISSUE THREAD

**Issue by ismell** [SublimeLinter/SublimeLinter#1061]
I just got the 4.0.1 upgrade and I randomly get the following:

```
SublimeLinter: #221 mylinter foo.txt ERROR:

 Linter crashed.

Traceback (most recent call last):
  File "lint.backend in .config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 97, in execute_lint_task
  File "lint.linter in .config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 797, in lint
  File "lint.linter in .config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 1139, in run
  File "lint.linter in .config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 1188, in communicate
  File "lint.util in .config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 225, in communicate
  File "./python3.3/subprocess.py", line 920, in communicate
  File "./python3.3/subprocess.py", line 1556, in _communicate
  File "./python3.3/subprocess.py", line 1670, in _communicate_with_poll
  File "./python3.3/subprocess.py", line 1607, in close_unregister_and_remove
OSError: [Errno 9] Bad file descriptor
```

My linter is accepting the buffer via stdin.

Running on linux with sublime 3157

**Comment by kaste** [comment]
Probably enable debug mode and open the console to get more info. What cmd do you actually run here?

**Comment by huochunpeng** [comment]
Me too. Same error on htmlhint. Debug mode doesn't provider any extra log.

**Comment by kaste** [comment]
I'm closing bc you're gone. 😥  We can re-open later.

**Comment by ismell** [comment]
I only work on the weekdays so I didn't respond during the weekend.

Enabling debug didn't show anything. I redacted some paths.

```
SublimeLinter: #684 SublimeLinter.sublime-package:782: 'glint' is linting 'ClientUtil.java'
SublimeLinter: #684 SublimeLinter.sublime-package:1178: glint: ClientUtil.java ['/usr/bin/glint', '--output=json', '--stdin', '/google/src/cloud/
SublimeLinter: #684 SublimeLinter.sublime-package:1181: glint: cwd: ...
SublimeLinter: #687 SublimeLinter.sublime-package:782: 'sublimesyntax' is linting 'ClientUtil.java'
SublimeLinter: #688 SublimeLinter.sublime-package:782: 'glint' is linting 'ClientUtil.java'
SublimeLinter: #688 SublimeLinter.sublime-package:1178: glint: ClientUtil.java ['/usr/bin/glint', '--output=json', '--stdin', '.../ClientUtil.java']
SublimeLinter: #688 SublimeLinter.sublime-package:1181: glint: cwd: ...
SublimeLinter: Buffer 31 inconsistent. Aborting lint. 
SublimeLinter: Buffer 31 inconsistent. Aborting lint. 
SublimeLinter: #689 SublimeLinter.sublime-package:782: 'sublimesyntax' is linting '<untitled 176>'
indexing [job 26]: spawning 1 workers to process 1 / 1 files
indexing [job 26]: indexed 1 files
SublimeLinter: #684 __init__.py:1273: ERROR: Linter crashed.

Traceback (most recent call last):
  File "lint.backend in ~/.config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 97, in execute_lint_task
  File "lint.linter in ~/.config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 797, in lint
  File "lint.linter in ~/.config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 1139, in run
  File "lint.linter in ~/.config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 1188, in communicate
  File "lint.util in ~/.config/sublime-text-3/Installed Packages/SublimeLinter.sublime-package", line 225, in communicate
  File "./python3.3/subprocess.py", line 920, in communicate
  File "./python3.3/subprocess.py", line 1556, in _communicate
  File "./python3.3/subprocess.py", line 1670, in _communicate_with_poll
  File "./python3.3/subprocess.py", line 1607, in close_unregister_and_remove
OSError: [Errno 9] Bad file descriptor
SublimeLinter: Buffer 31 inconsistent. Aborting lint. 
SublimeLinter: Buffer 31 inconsistent. Aborting lint. 
SublimeLinter: Buffer 31 inconsistent. Aborting lint. 
SublimeLinter: #688 SublimeLinter.sublime-package:809: glint output:
    {"files": []}
```

This is a custom linter. Could it be a race condition with the buffer being inconsistent and aborting the linter?

**Comment by braver** [comment]
Can you share the linter code, or at least where it interfaces with our “api”?

**Comment by ismell** [comment]
Here you go: https://pastebin.com/DmZQdrDM

**Comment by ismell** [comment]
I need to go back and analyze if `error_stream = util.STREAM_BOTH` is still required.

**Comment by kaste** [comment]
I don't find glint using google. 😑

**Comment by kaste** [comment]
But honestly, it just crashes as soon as we trigger it again. So it hates us for calling it parallel.

**Comment by ismell** [comment]
It's an internal only tool. I can take a look and see if there is any reason it can't be called in parallel. Is there a flag for sublimelinter to prevent it from dunning in parallel?

**Comment by kaste** [comment]
No there is no simple switch for that.

**Comment by kaste** [comment]
I mean you know python. You could override the run method and protect the super call with a lock. But then you hold the lock for the time your glint takes.

**Comment by ismell** [comment]
Not sure if this is a python bug or not, but pstree shows a prolonged running glint process:

```
        ├─sublime_text─┬─plugin_host─┬─glint
        │              │             └─4*[{plugin_host}]
        │              └─8*[{sublime_text}]
```

Doing a `ps` it shows it's defunct

```
me   18615  0.2  0.0      0     0 ?        Z    11:56   0:00 [glint] <defunct>
```

Something is not cleaning up dead processes correctly.

**Comment by kaste** [comment]
I thought you can actually debug glint. You maybe just take 'the' (which?) file descriptor away as soon as we run the program again. 

But without the Lock blah, you could also just eat the OSError.

**Comment by kaste** [comment]
Back to normal reasoning. 

- I cannot think of anything that glint could do that should cause a fd error here. I mean you close stdin or crash, and python should just handle it. 

- I cannot see what we do wrong. I mean we call `result = Popen(cmd, PIPE, PIPE, PIPE).communicate(source_code)`. We don't do low-level stuff. `communicate()` blocks anyway and it throws here.

However, I also cannot find any issues against python and `git blamed` the module. Essentially they had to rewrite it. Maybe there was a reason. 

However 2, if it's a race in python std library code, we should get more reports from users bc we throw that error in your face.

**Comment by kaste** [comment]
Re-open for visibility. 

We now have three users with this issue. All running OSX and linters accepting input on STDIN. 

@ismell Do you think that we *just* spawn too many threads? Do you know how I would catch exactly this error. Is it 

```
	except OSError as e: 
		if e.errno == 9: pass
```

**Comment by ismell** [comment]
I figured out how to reproduce it locally. I had an strace dump I was going through, but it was long. I don't actually have any theories yet. I'll try to do some more digging.

**Comment by ctf0** [comment]
am also having this god dam focken bug, is there a way to at least ignore it or dont open the console when its triggered ?

**Comment by pedrofoliveira** [comment]
For me it is also crashing on Windows...

Windows 7 Enterprise
Sublime Text Build 3143
Code in Ruby using Rubocop

SublimeLinter: #1 rubocop report_week.rb ERROR:

 Linter crashed.

Traceback (most recent call last):
  File "lint.backend in C:\Users\<my_name>\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 97, in execute_lint_task
  File "lint.linter in C:\Users\<my_name>\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 762, in lint
  File "lint.linter in C:\Users\<my_name>\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 1006, in run
  File "lint.linter in C:\Users\<my_name>\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 1055, in communicate
  File "lint.util in C:\Users\<my_name>\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 225, in communicate
  File "./python3.3/subprocess.py", line 920, in communicate
  File "./python3.3/subprocess.py", line 1206, in _communicate
BrokenPipeError: [Errno 32] Broken pipe

**Comment by kaste** [comment]
@pedrofoliveira Check file size and if it is not rather #347

**Comment by kaste** [comment]
Also checkout and test #1234 for reduced concurrency.

**Comment by pedrofoliveira** [comment]
The size of the file in the editor is definitely larger than 4kB (in this case it is about 7kB).
I have no idea about the cause so I don't know if it is just another instance of #347 or #1234... but nothing was crashing before updating to the latest version of SublimeLinter.

**Comment by ismell** [comment]
I think i found it. I was browsing the 3.3.6 source and they are using class variables and module variables. 

https://github.com/python/cpython/commit/580e007860e0142c9ea0eabec49560ca515e8476#diff-cc136486b4a8e112e64b57436a0619eb Uses a class variable???

https://github.com/python/cpython/blob/605765fa32f994778ae7d6c48945299228f111e5/Lib/subprocess.py#L468 Module variable....

There have been various fixes to communicate too: https://github.com/python/cpython/commits/3.4/Lib/subprocess.py

**Comment by tssajo** [comment]
Hi,

I'm having the exact same issue as @pedrofoliveira on Windows 7 x64.

I am trying to lint a Go ( aka. golang ) file. The file size is 4299 bytes.

Here is the error I get:

```
SublimeLinter: #2 gotype b2-pdf.go ERROR:
=========================================

Linter crashed.

Traceback (most recent call last):
  File "lint.backend in C:\Users\User\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 100, in execute_lint_task
  File "lint.linter in C:\Users\User\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 761, in lint
  File "lint.linter in C:\Users\User\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 1005, in run
  File "lint.linter in C:\Users\User\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 1054, in communicate
  File "lint.util in C:\Users\User\AppData\Roaming\Sublime Text 3\Installed Packages\SublimeLinter.sublime-package", line 226, in communicate
  File "./python3.3/subprocess.py", line 920, in communicate
  File "./python3.3/subprocess.py", line 1206, in _communicate
BrokenPipeError: [Errno 32] Broken pipe
```

I can reproduce this 100% of the time.

It worked before with an older version of SublimeLinter.
Only after a recent SublimeLinter update it started to crash like this.
I get this exact same error when I open other Go files as well.

I'm not sure if it is the same as https://github.com/SublimeLinter/SublimeLinter/issues/347 but if it is then please reopen that issue.

Thanks!

**Comment by braver** [comment]
@kaste do you know what’s the status here? Should our new features to limit concurrency fix it?

**Comment by braver** [comment]
Well at least by now the topic has diverged from the title and this is now a duplicate of #1327, so let’s continue there is needed.

**Comment by kaste** [comment]
We have one more report https://github.com/SublimeLinter/SublimeLinter-rubocop/issues/57 also OSX. That's 4 users in 2 months.

**Comment by kaste** [comment]
I provided a patch against core https://github.com/SublimeTextIssues/Core/issues/2293#issuecomment-391642348

Now, as @ismell pointed out, we have global state here, so if one plugin pollutes this state SublimeLinter or any other thing that uses subprocess can crash. I could reproduce this in that way. The patch is against the core `exec.py` which is probably highly used within SublimeText, but honestly, every plugin can, as I said, pollute python here, and I still don't see that we do.

**Comment by ismell** [comment]
The core problem is in pythons exec() method. @wbond, not sure if it's possible to uprev python, or cherry pick the latest communicate.py.

**Comment by jasjuang** [comment]
I ran into the same problem for my [CMakeFormat ](https://packagecontrol.io/packages/CMakeFormat) plugin. When I run it on a Windows and on a large file the below error happens

```
['cmake-format.exe', '--line-width', '80', '--tab-size', '2', '--max-subargs-per-line', '3', '--separate-ctrl-name-with-space', 'False', '--separate-fn-name-with-space', 'False', '--dangle-parens', 'False', '--bullet-char', '*', '--enum-char', '.', '--line-ending', 'unix', '--command-case', 'lower', 'C:\\Users\\jasju\\Desktop\\Camera\\CMakeLists.txt']
Traceback (most recent call last):
  File "C:\Program Files\Sublime Text 3\sublime_plugin.py", line 1072, in run_
    return self.run(edit)
  File "C:\Users\jasju\AppData\Roaming\Sublime Text 3\Installed Packages\CMakeFormat.sublime-package\cmake_format.py", line 212, in run
  File "./python3.3/subprocess.py", line 920, in communicate
  File "./python3.3/subprocess.py", line 1206, in _communicate
BrokenPipeError: [Errno 32] Broken pipe
``` 

I am able to run it successfully on the same file on Ubuntu, and if I am able to run it successfully on a smaller file on Windows too.

**Comment by kaste** [comment]
@jasjuang That's unrelated, you don't use SublimeLinter here. As a hint, don't pipe data through stdin/communicate if the receiving program does not open/read from it. In your traceback you tell cmake-format to operate on a file on disk not on stdin.

**Comment by jasjuang** [comment]
@kaste It's not sublimelinter but the error message is very similar. I suspect the root cause is the same. cmake-format is supposed to operate on a file on disk without a problem.

**Comment by kaste** [comment]
If you operate on a file on disk, cmake-format will not open stdin and you must not use communicate/pipe in data. That's what I'm saying.

**Comment by jasjuang** [comment]
@kaste I don't fully understand your point. If cmake-format will not open stdin if it's operating on a file on disk, shouldn't it fail on all scenarios? (Ubuntu, smaller files on Windows)

**Comment by kaste** [comment]
Buffer size. How large is your 'larger' file?

**Comment by jasjuang** [comment]
@kaste It's 136 lines only, and the file size is 6KB.
